### PR TITLE
[Feature] Auto-staking + add short_reward_weight to PoolInfoResponse

### PR DESCRIPTION
**Pull request by csanti**
### Auto-staking

We add the possibility of Providing liquidity and directly staking the minted LP tokens to the staking contract.

To do so, staking contract needs to be the receiver of the LP tokens. Currently there is no way to change the receiver of Terraswap's `ProvideLiquidity` oparation; therefore, staking contract needs to provide liquidity for the user, acting as a relay.

The operation is implemented following this flow: 

_User_
1. User increases allowance of token for staking contract
2. User executes AutoStake operation sending the native asset coin

_Staking contract (AutoStaking)_
1. Transfer token asset to staking contract
2. Increase allowance of token for pair contract
3. Provide liquidity
4. Execute staking hook

_Staking contract (AutoStakingHook)_
1. Check how many LP tokens received
2. Increase bond amount for the user

### Modify PoolInfoResponse

`short_reward_weight` is added to `PoolInfoResponse`.

To do that, also modified `short_reward_weight` calculation so that it is only updated at the same time as the `premium`.

### CODE REVIEW

**Comment by csanti** [comment]
- Added a fix to deduct tax from native asset when relaying the provide liquidity msg.
- Also fixed wrong asset address when querying staking token balance.

File: contracts/mirror_staking/src/staking.rs [hunk 1]
```diff
@@ -148,6 +152,134 @@ pub fn decrease_short_token<S: Storage, A: Api, Q: Querier>(
...
+                        Asset {
+                            amount: token_amount,
+                            info: AssetInfo::Token {
+                                contract_addr: token_addr.clone(),
+                            },
+                        },
+                    ],
+                    slippage_tolerance,
+                })?,
+                send: vec![native_asset.deduct_tax(deps)?],
```
Comment (YunSuk-Yeo) [inline contracts/mirror_staking/src/staking.rs:236]: `deduct_tax` conduct two queries `tax_rate` & `tax_cap`.
We can skip this step by subtracting the pre-computed tax_amount like L224.

File: contracts/mirror_staking/src/staking.rs [hunk 2]
```diff
@@ -148,6 +152,134 @@ pub fn decrease_short_token<S: Storage, A: Api, Q: Querier>(
...
+            }
+        }
+    }
+
+    // will fail if one of them is missing
+    let native_asset: Asset = match native_asset_op {
+        Some(v) => v,
+        None => return Err(StdError::generic_err("Missing native asset")),
+    };
+    let (token_addr, token_amount) = match token_info_op {
```
Comment (YunSuk-Yeo) [inline contracts/mirror_staking/src/staking.rs:183]: maybe good to check the token is in the asset pool store

File: contracts/mirror_staking/src/staking.rs [hunk 3]
```diff
@@ -148,6 +152,134 @@ pub fn decrease_short_token<S: Storage, A: Api, Q: Querier>(
...
+    // assert staking token
+    let pool_info: PoolInfo =
+        read_pool_info(&deps.storage, &deps.api.canonical_address(&asset_token)?)?;
+
+    if pool_info.staking_token != deps.api.canonical_address(&staking_token)? {
+        return Err(StdError::generic_err("Invalid staking token"));
+    }
+
+    // stake all lp tokens received
+    let staking_token_amount = query_token_balance(&deps, &staking_token, &env.contract.address)?;
```
Comment (YunSuk-Yeo) [inline contracts/mirror_staking/src/staking.rs:278]: We need to pass previous balance and subtract that from the current balance to get actual staking amount

**Review by YunSuk-Yeo — CHANGES_REQUESTED** [review]

File: contracts/mirror_staking/src/staking.rs (hunk 1 above)
Comment (csanti) [reply contracts/mirror_staking/src/staking.rs]: changed it!

File: contracts/mirror_staking/src/staking.rs (hunk 2 above)
Comment (csanti) [reply contracts/mirror_staking/src/staking.rs:183]: I had that check in the AutStakeHook, but moved it to AutoStake operation, I agree it makes more sense to check before.

File: contracts/mirror_staking/src/staking.rs (hunk 3 above)
Comment (csanti) [reply contracts/mirror_staking/src/staking.rs]: Fixed it

**Comment by csanti** [comment]
Implemented requested changes, @YunSuk-Yeo please review again

**Review by YunSuk-Yeo — APPROVED** [review]
LGTM

**Pull request merged**